### ISSUE TITLE
Page refresh firing before saveChat

### ISSUE THREAD

**Issue by Godrules500** [vercel/ai-chatbot#302]
I have a situation every now and then when I create a new chat that, and the saveChat takes a little bit of time, the chat.tsx router.refresh() fires too soon. How can I ensure that it waits for saveChat to finish?

This is the code that is firing before the saveChat finishes. So the assistant responds, and the saveChat finishes after the aiState.done is called.
`  useEffect(() => {
    const messagesLength = aiState.messages?.length
    if (messagesLength === 2) {
      router.refresh()
    }
  }, [aiState.messages, router])`

**Comment by amxv** [comment]
having the same issue atm, were you able to fix it?

**Comment by Godrules500** [comment]
Hey @amxv yes, I was able to resolve it.  Since I'm using gemini, I had to follow the AI SDK 3 example "next-ai-rsc". But essentially I refactored a few things. 

Changes:

- Added a "saved" property to AIState (to prevent saving multiple times).
- Refactored saveChat called inside of unstable_onSetAIState to a method
- Called await saveChat when text completion "isFinal" was true.

I now call saveChat inside of unstable_onSetAIState  like this
```
unstable_onSetAIState: async ({ state, done }) => {
    'use server'

    const session = await auth()

    if (session && session.user) {

      // only save it once it's done. No need to save it after each streamed result
      if (done && !state.saved) {
        await addOrUpdateChat(state, session)
        state.saved = true
      }
    } else {
      return
    }
  }
})
```

I then call await saveChat(state) when isFinal is true and set saved to true.
```
let state = aiState.get()
        state.messages = [
          ...state.messages,
          {
            id: nanoid(),
            role: 'assistant',
            content
          }
        ]

        await saveChat(state)

        reply.done()
        aiState.done({ ...state, done: isFinal, saved: true })
```

And then the saveChat() is this

```
async function saveChat(state: AIState, session: Session | null = null) {
  if (!session) {
    session = await auth()
    if (!session?.user?.id) {
      throw new Error('User not Authorized to save')
    }
  }

  const { chatId, messages } = state

  const createdAt = Date.now()
  const userId = session.user?.id as string
  const path = `/chat/${chatId}`
  const title = messages[0].content.substring(0, 100)

  const chat: Chat = {
    id: chatId,
    title,
    userId,
    createdAt,
    path,
    messages
  }

  await dynamoDbService.saveChat(chat)
}
```

Let me know if that fixes it for you too or if you run into any issues with it please!

**Comment by athrael-soju** [comment]
@Godrules500 could you please raise a PR with the fix? 

Fixes such as this one are super helpful for keeping the template updated.

**Comment by audvin** [comment]
Experiencing the same issue.

**Comment by Godrules500** [comment]
@athrael-soju  I will try to get to it soon!

I also found another solution to another issue if it helps y'all.
If you go to new chat, and then clear history, and then send a prompt, it will refresh because the chat ids are out of sync. My fix for me is as follow.

on chat.tsx, this is the problem area for me. 
```
useEffect(() => {
    setNewChatId(id)
  })
```
I updated mine to 
```
  useEffect(() => {
    // This is here, because it resolve a bug when you go to 'new chat' --> clear history, and then generate a new chat, the IDs were out of sync.
    updateAIState({ ...aiState, chatId: id! })
    setNewChatId(id)
  })
```

then on actions.tsx, I added this.
```
async function updateAIState(currentAIState: any) {
  'use server'
  const aiState = getMutableAIState<typeof AI>()
  aiState.done({ ...currentAIState })
}
```
don't forget to add it where submitUserMessage is added.

Then update chat.tsx to

**Comment by yaberkane05** [comment]
Hi, any news on this issue ?

**Comment by navkuun** [comment]
Made the changes as said here, but still facing this issue. Anyone else?

My current code:

```
export type AIState = {
  chatId: string
  messages: Message[]
  saved?: boolean
}

export type UIState = {
  id: string
  display: React.ReactNode
}[]


async function addOrUpdateChat(state: AIState, session: Session | null = null) {
  if (!session) {
    session = await auth()
    if (!session?.user?.id) {
      throw new Error('User not Authorized to save')
    }
  }

  const { chatId, messages } = state

  const createdAt = new Date()
  const userId = session?.user?.id as string
  const path = `/chat/${chatId}`

  const firstMessageContent = messages[0].content as string
  const title = firstMessageContent.substring(0, 100)

  const flowchart = await getFlowchart(chatId)

  const chat: Chat = {
    id: chatId,
    title,
    userId,
    createdAt,
    path,
    messages,
    flowchart: flowchart || { nodes: [], edges: [] } // Use existing flowchart data if available, otherwise initialize with empty arrays
  }

  await saveChat(chat)
}

export const AI = createAI<AIState, UIState>({
  actions: {
    submitUserMessage
  },
  initialUIState: [],
  initialAIState: { chatId: nanoid(), messages: [] },
  onGetUIState: async () => {
    'use server'

    const session = await auth()

    if (session && session.user) {
      const aiState = getAIState()

      if (aiState) {
        const uiState = getUIStateFromAIState(aiState)
        return uiState
      }
    } else {
      return
    }
  },
  onSetAIState: async ({ state, done }) => {
    'use server'

    const session = await auth()

    if (session && session.user) {
      // only save it once it's done. No need to save it after each streamed result
      if (done && !state.saved) {
        await addOrUpdateChat(state, session)
        state.saved = true
      }
    } else {
      return
    }
  }
})
```

**Comment by yaberkane05** [comment]
same for me.

**Comment by athrael-soju** [comment]
I'd wait a bit for the next version of the template to come out, as this one still has some breaking issues.

**Comment by navkuun** [comment]
any fix to this?

**Comment by yorickvanzweeden** [comment]
**TL;DR: Do not call aiState.done multiple times for the same action**

What I figured out, is the following:

- Client and server correspond using deltas
- State in the server is stored using AsyncLocalStorage
- aiState.update updates the current state and calls onSetAIState. (No updates are sent yet)
- aiState.done does the same as aiState.update AND it computes a diff that will be sent.

If you call aiState.done multiple times, only the first diff will be sent. In my case, the AIState was not correctly in sync and the refresh occurred too fast

**Comment by Saran33** [comment]
@Godrules500 nice solution, thank you🚀 I incorporated some of those changes into a branch but I don't have permissions to make a PR here. I forked the repo if anyone wants to apply those changes here https://github.com/vercel/ai-chatbot/pull/399

**Comment by Saran33** [comment]
This doesn't appear to be an issue on the demo site though so it may have already been resolved by the Vercel team but the changes not applied to this repo. I said I'd add the changes here anyway in any case it's useful to anyone.

**Comment by preshetin** [comment]
Thanks @Saran33 - I used the code from your #399 PR and it looks like it does fix that!
